### PR TITLE
v6: lerna, deps, react fix

### LINKED ISSUE #44: Use Lerna
I did a little feasibility test to see if we can use [Lerna](https://lernajs.io/) for handling our two packages.

Turns out [Lerna doesn't support Yarn](https://github.com/lerna/lerna/issues/371) yet. Interestingly Jest from Facebook (the company who made Yarn on the first place [is still using Lerna](https://github.com/facebook/jest/blob/master/lerna.json) *v2.0.0-beta.31* despite using [yarn](https://github.com/facebook/jest/blob/master/yarn.lock)).

Anyway, [there are a bunch of yarn issues](https://github.com/lerna/lerna/issues?utf8=%E2%9C%93&q=yarn) on Lerna that one has to investigate.

My two cents:

* Use the `package.json` in the root for now until we have enough packages that justifies bringing another dependency into the project
* Wait and see if Lerna supports Yarn fully or [Yarn will swallow Lerna](https://github.com/yarnpkg/yarn/issues/946#issuecomment-264597575). Either way, it seems to early to decide.

### PR DESCRIPTION
**Pull request by nosilleg**
Not separate commits, sorry.

Fixed eslint-config-schibsted-react because I borked the SPT merge and missed a dependency.
Update some dependencies including unicorn major bump to V3.
Add Lerna to resolve #44

### CODE REVIEW

File: lerna.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "lerna": "2.5.1",
+  "version": "6.0.0",
```
Comment (SimenB) [inline lerna.json:3]: could use independent versioning, but not biggie

**Review by SimenB — APPROVED** [review]

File: lerna.json (hunk 1 above)
Comment (nosilleg) [reply lerna.json:3]: Was intentional to not use independent.  😄 

If you have some good reading on why independent is preferable, I'm happy to read and reevaluate on the next release.

Comment (SimenB) [reply lerna.json:3]: Short of it is that even though you make a breaking change in the react config, there's no need to make a breaking change in all of the other packages

Comment (nosilleg) [reply lerna.json:3]: Understood.  That was a consideration when choosing but my preference is fixed versions.  Inspecting a package.json it's easy to tell when you have compatible versions.  This isn't an issue if you're on the latest versions, but if you're a few behind you have more effort to look up which package goes with which.

**Comment by nosilleg** [comment]
@SimenB I've updated the dependencies as we discussed, but I've removed all peerDependencies in favour of pure dependency usage.

File: package.json [hunk 2]
```diff
@@ -25,7 +21,8 @@
   },
   "devDependencies": {
     "eslint-config-schibsted": "file:./packages/eslint-config-schibsted",
-    "eslint-config-schibsted-node": "file:./packages/eslint-config-schibsted-node"
+    "eslint-config-schibsted-node": "file:./packages/eslint-config-schibsted-node",
```
Comment (SimenB) [inline package.json]: you shouldn't need these as yarn hoists them up automatically

File: packages/eslint-config-schibsted-chai/yarn.lock [hunk 3]
```diff
@@ -12,15 +12,15 @@ acorn@^3.0.4:
   version "3.3.0"
```
Comment (SimenB) [inline packages/eslint-config-schibsted-chai/yarn.lock]: delete nested yarn lockfiles

**Review by SimenB — CHANGES_REQUESTED** [review]
`rm packages/*/yarn.lock`, then we should be gtg

File: package.json (hunk 2 above)
Comment (nosilleg) [reply package.json]: Ah yes, with the workspaces. Patch incoming.

File: packages/eslint-config-schibsted-chai/yarn.lock (hunk 3 above)
Comment (nosilleg) [reply packages/eslint-config-schibsted-chai/yarn.lock]: I would have expected that to be cleaned up automatically.

Done.

**Review by SimenB — APPROVED** [review]

File: packages/eslint-config-schibsted-chai/package.json [hunk 4]
```diff
@@ -9,11 +9,8 @@
     "test": "eslint ."
   },
   "dependencies": {
-    "eslint": "^4.10.0",
-    "eslint-plugin-chai-friendly": "^0.4.0"
-  },
-  "peerDependencies": {
-    "eslint": "^4.10.0"
+    "eslint": "^4.13.1",
```
Comment (SimenB) [inline packages/eslint-config-schibsted-chai/package.json]: I wonder if `eslint` itself should be a peer, but not the configs and plugins?

Comment (nosilleg) [reply packages/eslint-config-schibsted-chai/package.json]: It would gain is a warning when installing, but I don't think it improves the resolution process. As we discussed originally my plan was to leave eslint as the only peer, but then I thought maybe it didn't deserve to be special cased since any package version conflicts will give unpredictable outcomes.

This method also means that eslint doesn't have to be specified at the root.  How likely that is I'm unsure since I don't know how others will mix this with other linting packages.

When you go to update the proxy config for Finn, will the peer warning  for just eslint be sufficiently beneficial?

Comment (SimenB) [reply packages/eslint-config-schibsted-chai/package.json]: This is about allowing consumers to control the version of eslint independently from the configs

>any package version conflicts will give unpredictable outcomes.

how so?

>When you go to update the proxy config for Finn, will the peer warning for just eslint be sufficiently beneficial?

yes

Comment (nosilleg) [reply packages/eslint-config-schibsted-chai/package.json]: Correct my understanding where it's wrong.

Consumers can control the version of any package by using the `resolutions` section of `package.json`.  

If a eslint config/plugin has multiple versions installed, the version at the root will be the version used when eslint executes.  The version in the root might not be the version that we wanted.  I believe this would be the reasoning for AirBnB to make all eslint plugins/configs peerDependencies.

You said yes to the warning about eslint, and only eslint, would provide benefit.  Given the above, can you please clarify why?  I would have assumed that warnings for all or warnings for none would be the preferable approach.

I could go either way.  

I originally had a lot more in here about how peerDependency installs could go wrong, but that just lead me to thinking that specifying peerDependencies is the better option.  Packages like ESLint that require a CLI to read from the root `node_modules` seem like the poster package for peerDependencies.  It doesn't stop version conflicts, but it does expose the warnings.

But...

The warnings shortcut debugging, but I'm getting more blind to them with the recent bug(s) with false positives.  And there will be situations where not all of you dependencies have updated to the latest eslint/config/plugin.  So you'll either have to choose to not update or see the error and get unexpected outcomes anyway.

Plus not having to install things in the root that I'm not actually using myself is nice.

In the case of eslint-config-finn, you're going to have a dependency that is 3 levels removed from the root `package.json`.  Do we really want a peerDependency chain so far removed from its use?

eslint-config-finn -> eslint-config-schibsted -> eslint-config-airbnb -> eslint-plugin-import

These will need to be installed at the root, with some appearing unused:
eslint
eslint-plugin-import
eslint-plugin-jsx-a11y
eslint-plugin-react
eslint-plugin-chai-friendly
eslint-plugin-flowtype
eslint-config-airbnb-base
eslint-config-airbnb
eslint-config-prettier
eslint-plugin-prettier
eslint-plugin-unicorn
prettier

... and then...

I've just spotted some `>=` versioning, so even if we get rid of all the `unmet peerDependency` messages, we might still be getting unexpected behavior because of changes in major versions.

After all that I'm actually leaning towards making them all peerDependencies.  Hopefully I've gotten something wrong and you can talk me out of it.  😁

Comment (SimenB) [reply packages/eslint-config-schibsted-chai/package.json]: Here's an example of how having it as `dependencies` is a potential footgun, and why peer deps might be preferred: https://github.com/eslint/eslint/issues/8547

Because of semver, the only way to ensure a single version is really to use peer deps, but it's annoying. See https://github.com/eslint/eslint/issues/3458

I think maybe `plugins` should be peer deps, while `config` can be deps. (in practice)

Eslint itself should be peer. And prettier should be peer.

Agreed?

Comment (nosilleg) [reply packages/eslint-config-schibsted-chai/package.json]: Thanks for doing the leg work finding those threads. It's a shame people discovered these issues 2 years ago and it still isn't resolved.

After reading those threads and a lot of tangents is confirmed my assertions above that the benefit of using peerDependencies is the warning message on install. It doesn't prevent version conflicts.

Anyway, I've updated the code to use peerDependencies, and corrected the use of configs so that they are at least the correct version. This can't be done nicely for configs that come from plugins, so I have just given in that there may be plugin/config mismatches as well as plugin mismatches.

There was at least one extra peerDependency that was required, but want listed as a peer in the dependency. Hopefully I've covered the lot.

READMEs updated to reflect the requirements.

**Comment by nosilleg** [comment]
Please give another review with the latest changes. I'll squash on merge.

File: packages/eslint-config-schibsted-chai/index.js [hunk 5]
```diff
@@ -4,5 +4,5 @@ module.exports = {
   env: {
     node: true
   },
-  extends: ["./chai.js"].map(require.resolve)
```
Comment (SimenB) [inline packages/eslint-config-schibsted-chai/index.js:7]: why this change?

Comment (SimenB) [reply packages/eslint-config-schibsted-chai/index.js:7]: then `eslint` will just see `./chai.js` and will have to figure out the path. simpler and clearer to give absolute path IMO

**Review by SimenB — APPROVED** [review]
Just a nit, this LGTM!

File: packages/eslint-config-schibsted-chai/index.js (hunk 5 above)
Comment (nosilleg) [reply packages/eslint-config-schibsted-chai/index.js:7]: The main benefit of using the require.resolve is that it forces resolution to the packages specified in this packages package.json instead of the package that eslint finds in the root node_modules. This isn't a concern for the relative path since it's relative to this package. (Although this wasn't one of the tests I did.)

I also made the call that having less code here was better. I don't know the speed differences of resolving the path here, versus it being resolved somewhere later, but I'm presume it's negligible in this use case.

Happy to revisit, but will merge as-is since I have your approval.
